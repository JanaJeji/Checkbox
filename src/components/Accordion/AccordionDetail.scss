.accordion-detail-container{
  border: 1px solid #ccc;
  border-top: none;
  transition: height 0.2s ease-in-out;
  overflow: hidden;

}
.accordion-detail-container-content{
  padding: 15px 20px;
}
