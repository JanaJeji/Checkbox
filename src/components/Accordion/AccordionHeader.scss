.accordion-header{
  border: 1px solid #ccc;
  cursor: pointer;
  user-select: none;


}
.accordion-header:not(:first-of-type) {
  border-top: 0;
}
.accordion-header-title{
  width: 100%;
  margin: 0;
}

.accordion-header-btn{
  width: 100%;
  display: flex;
  align-items: center;
  border: 0;
  font-size: 22px;
  font-weight: 500;
  background-color: #fff;
}

.accordion-header-btn > p {
  width: 100%;
  height: 100%;

  padding-left: 15px;
  padding-top: 15px;
}

.accordion-header-btn::after{
  content: '';
  flex-shrink: 0;
  width: 22px;
  height: 22px;
  margin-left: auto;
  background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16' fill='%23212529'%3e%3cpath fill-rule='evenodd' d='M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z'/%3e%3c/svg%3e");
  background-repeat: no-repeat;
  background-size: 18px;
  transition: transform 0.2s ease-in-out;
}

.accordion-header.active .accordion-header-btn::after {
  transform: rotate(-180deg);
}

