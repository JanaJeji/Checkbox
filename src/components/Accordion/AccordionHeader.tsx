import React from 'react';

export type AccordionHeaderProps = {
  children: JSX.Element | JSX.Element[];
  btnOnClick?: () => void;
  isOpen?: boolean;
};
const AccordionHeader = (props: AccordionHeaderProps) => {
  const childArray = Object.values(props.children);
  return (
    <div
      onClick={props.btnOnClick}
      className={`accordion-header ${props.isOpen ? 'active' : ''}`}
    >
      {Array.isArray(props.children) ? (
        <>
          <div className="accordion-header-title">{childArray[0]}</div>
          <div className="accordion-header-btn">
            <p style={{ color: '#6c757ded', fontSize: '18px' }}>
              {childArray.slice(1).map((elem, idx) => {
                return elem.props.children;
              })}
            </p>
          </div>
        </>
      ) : (
        <div className="accordion-header-btn">{props.children}</div>
      )}
    </div>
  );
};

export default AccordionHeader;
