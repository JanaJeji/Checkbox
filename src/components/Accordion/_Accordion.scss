
.accordion-header{
  border-bottom: 0.5px solid $gray-400;
}
.accordion-header-title {
  width: 100%;
  margin: 0;
}

.accordion-header-btn {
  width: 100%;
  display: flex;
  align-items: center;
  font-size: 22px;
  font-weight: 500;

}

.accordion-header-btn > p {
  padding-left: 20px;
  padding-top: 15px;
}

.accordion-header-btn::after {
  content: '';
  flex-shrink: 0;
  width: 22px;
  height: 22px;
  margin-left: auto;
  margin-right: 10px;
  background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16' fill='%23212529'%3e%3cpath fill-rule='evenodd' d='M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z'/%3e%3c/svg%3e");
  background-repeat: no-repeat;
  background-size: 22px;
  transition: transform 0.2s ease-in-out;
}

.accordion-header.active .accordion-header-btn::after {
  transform: rotate(-180deg);
}

.accordion-container {
  border: 1px solid $gray-400;
  border-bottom: 0.5px;
  cursor: pointer;
  user-select: none;
  background-color: $white;
  &.disabled,
  &[disabled] {
    cursor: not-allowed;
    opacity: $btn-disabled-opacity;
    box-shadow: none;
    background-color: $gray-600;
  }
}
.accordion-container.active {
  border: 1px solid $gray-400;
  margin: 16px 0;
  .accordion-header{
    border-bottom: 0;
  }
}
.accordion-container:last-of-type {
  border-bottom: 1px solid $gray-400;
}

.accordion-detail-container {

  transition: height 0.2s ease-in-out;
  overflow: hidden;
}

.accordion-detail-container-content {
  padding: 15px 20px;
}
