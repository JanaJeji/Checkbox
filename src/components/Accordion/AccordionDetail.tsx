import React, { useRef, useState, useEffect } from "react";
import "./AccordionDetail.scss";


type AccordionDetailProps = {
  children: JSX.Element;
  isOpen?: boolean;
};
function AccordionDetail(props: AccordionDetailProps) {
  const [height, setHeight] = useState(0);
  const detailRef = useRef<HTMLDivElement>(null);
  useEffect(() => {
    if (props.isOpen) {
      const detailEl = detailRef.current as HTMLDivElement;
      console.log(detailEl?.scrollHeight);
      setHeight(detailEl?.scrollHeight);
    } else {
      console.log("elsedyizz");
      setHeight(0);
    }
  }, [props.isOpen]);
  return (
    <div
      ref={detailRef}
      style={{ height }}
      className="accordion-detail-container"
    >
      <div className="accordion-detail-container-content">{props.children}</div>
    </div>
  );
}

export default AccordionDetail;
