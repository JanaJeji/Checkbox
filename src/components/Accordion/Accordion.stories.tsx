import React from 'react';
import { action } from '@storybook/addon-actions';
import Accordion from './Accordion';
import AccordionDetail from './AccordionDetail';
import AccordionHeader from './AccordionHeader';

export default {
  title: 'Accordion',
  component: Accordion,
};

export const DefaultAccordion = () => {
  return  <Accordion>
            <AccordionHeader>
              <p>hello</p>
            </AccordionHeader>
            <AccordionDetail>
              <p>New Accoirdon detail helloooo</p>
            </AccordionDetail>
          </Accordion>
}

export const MultipleAccordions = () => {
  return  <><Accordion>
            <AccordionHeader>
              <p>hello</p>
            </AccordionHeader>
            <AccordionDetail>
              <p>New Accoirdon detail helloooo</p>
            </AccordionDetail>
          </Accordion>
          <Accordion>
          <AccordionHeader>
            <p>hello</p>
          </AccordionHeader>
          <AccordionDetail>
            <p>New Accoirdon detail helloooo</p>
          </AccordionDetail>
        </Accordion>
        <Accordion>
        <AccordionHeader>
          <p>hello</p>
        </AccordionHeader>
        <AccordionDetail>
          <p>New Accoirdon detail helloooo</p>
        </AccordionDetail>
      </Accordion>
      </>
}


export const DisabledAccordion = () => {
  return  <><Accordion>
            <AccordionHeader>
              <p>hello</p>
            </AccordionHeader>
            <AccordionDetail>
              <p>New Accoirdon detail helloooo</p>
            </AccordionDetail>
          </Accordion>
          <Accordion>
          <AccordionHeader>
            <p>hello</p>
          </AccordionHeader>
          <AccordionDetail>
            <p>New Accoirdon detail helloooo</p>
          </AccordionDetail>
        </Accordion>
        <Accordion disabled>
        <AccordionHeader>
          <p>hello</p>
        </AccordionHeader>
        <AccordionDetail>
          <p>New Accoirdon detail helloooo</p>
        </AccordionDetail>
      </Accordion>
      </>
}

