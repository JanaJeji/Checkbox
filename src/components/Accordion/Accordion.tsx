import React, { useState, InputHTMLAttributes, useEffect } from 'react';
import AccordionDetail from './AccordionDetail';
import AccordionHeader from './AccordionHeader';
import { classNames } from '../../utils/classNames';

export interface IAccordionProps {
  children: JSX.Element[];
  disabled?: boolean;
  expanded?: boolean;
  sx?: {};
  rounded?: boolean;
  onClick?: (a?: boolean) => void;

}

export type bhdrProps = InputHTMLAttributes<HTMLInputElement>;
type InputArgs = IAccordionProps & Omit<bhdrProps, keyof IAccordionProps>;
function Accordion(props: InputArgs) {
  const { children, disabled, onClick, expanded, sx, ...rest } = props;

  const [open, setOpen] = useState<boolean>(false);
  useEffect(() => {
    setOpen(expanded as boolean);
  }, [expanded]);

  const btnOnClick = (a: boolean) => {
    if (!disabled) {
      setOpen(!a);
    }
  };

  const checkOpenorNot = () => {
    if (typeof onClick === 'function') {
      onClick?.(!open);
    }
  };
  let styleClasses = classNames('accordion-container', {
    disabled: !!disabled,
  });

  return (
    <div
      onClick={checkOpenorNot}
      className={`${styleClasses} ${expanded || open ? 'active' : ''}`}
      style={sx}
    >
      {children.map((elem, idx) => {
        if (elem.type.name === 'AccordionHeader') {
          return (
            <AccordionHeader
              isOpen={open}
              btnOnClick={() => btnOnClick(open)}
              key={`${elem.type.name}${idx}`} //uid
            >
              {elem.props.children}
            </AccordionHeader>
          );
        } else {
          return (
            <AccordionDetail  expanded={props.expanded} isOpen={open} key={idx}>
              {elem.props.children}
            </AccordionDetail>
          );
        }
      })}
    </div>
  );
}

export default Accordion;
