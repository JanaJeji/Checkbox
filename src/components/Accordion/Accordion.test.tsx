import React from 'react';
import { render, fireEvent,screen } from '@testing-library/react';
import Accordion from "./Accordion";

import AccordionHeader from "./AccordionHeader";
import AccordionDetail from "./AccordionDetail";

describe('Accordion', () => {

  it('should match snapshot', () => {
    const { asFragment } = render(
    <Accordion>
    <AccordionHeader >
      <p>hello</p>
    </AccordionHeader>
    <AccordionDetail>
      <p>New Accoirdon detail helloooo</p>
    </AccordionDetail>
  </Accordion>)
  expect(asFragment()).toMatchSnapshot();
  });

  it('should render items', () => {
    const headerSec = 'HelloFromHeader'
    const detailSec = 'HelloFromDetail'
    render(
      <Accordion>
      <AccordionHeader >
        <p>{headerSec}</p>
      </AccordionHeader>
      <AccordionDetail>
        <p>{detailSec}</p>
      </AccordionDetail>
    </Accordion>)
    const headerEl = screen.queryByText(headerSec)
    const detailEl = screen.queryByText(detailSec)
    expect(headerEl).toBeInTheDocument();
    expect(detailEl).toBeInTheDocument();
  });

  it('should open when the expanded props passed', () => {
    const headerSec = 'HelloFromHeader'
    const detailSec = 'HelloFromDetail'
    render(
      <Accordion expanded>
      <AccordionHeader >
        <p>{headerSec}</p>
      </AccordionHeader>
      <AccordionDetail>
        <p>{detailSec}</p>
      </AccordionDetail>
    </Accordion>)
    const activeDiv = document.querySelector('.accordion-container.active')
    expect(activeDiv).toHaveClass('accordion-container active')
  })

  it('should open when the header clicked', () => {
    const headerSec = 'HelloFromHeader'
    const detailSec = 'HelloFromDetail'
    render(
      <Accordion>
      <AccordionHeader >
        <p>{headerSec}</p>
      </AccordionHeader>
      <AccordionDetail>
        <p>{detailSec}</p>
      </AccordionDetail>
    </Accordion>)
    const headerEl = screen.queryByText(headerSec) as HTMLDivElement
    fireEvent.click(headerEl)
    const activeDiv = document.querySelector('.accordion-container.active')
    expect(activeDiv).toHaveClass('accordion-container active')
  })
  it('should close when the accordion already open', () => {
    const headerSec = 'HelloFromHeader'
    const detailSec = 'HelloFromDetail'
    render(
      <Accordion>
      <AccordionHeader >
        <p>{headerSec}</p>
      </AccordionHeader>
      <AccordionDetail>
        <p>{detailSec}</p>
      </AccordionDetail>
    </Accordion>)
    const headerEl = screen.queryByText(headerSec) as HTMLDivElement
    const currentDiv = headerEl.closest('.accordion-container')

    fireEvent.click(headerEl)
    fireEvent.click(headerEl)
    expect(currentDiv).not.toHaveClass('active')
  })
  it('should have a style when props passed', () => {
    const headerSec = 'HelloFromHeader'
    const detailSec = 'HelloFromDetail'
    render(
      <Accordion sx={{backgroundColor : 'red'}}>
      <AccordionHeader >
        <p>{headerSec}</p>
      </AccordionHeader>
      <AccordionDetail>
        <p>{detailSec}</p>
      </AccordionDetail>
    </Accordion>)

    const accordionDiv = document.querySelector('.accordion-container')
    expect(accordionDiv).toHaveStyle('background-color : red;')
  })
  it('should have a render multiple child elements', () => {
    const headerSec = 'HelloFromHeader'
    const headerSec2 = 'HelloFromHeader2'
    const detailSec = 'HelloFromDetail'
    render(
      <Accordion sx={{backgroundColor : 'red'}}>
      <AccordionHeader >
        <p>{headerSec}</p>
        <p>{headerSec2}</p>
        <p>{headerSec2}</p>
        <p>{headerSec2}</p>
      </AccordionHeader>
      <AccordionDetail>
        <p>{detailSec}</p>
      </AccordionDetail>
    </Accordion>)

    const accordionHeaderDiv = document.querySelector('.accordion-header-btn') as HTMLDivElement

    expect(Object.values(accordionHeaderDiv)).toHaveLength(3)
  })
  // it('should open one at a time', () => {
  //   const headerSec = 'HelloFromHeader'
  //   const detailSec = 'HelloFromDetail'
  //   const headerSec2 = 'HelloFromHeader'
  //   const detailSec2 = 'HelloFromDetail'
  //   const [expanded, setExpanded] = React.useState<string | false >(false);

  // const handleChange = (panel : string) => {
  //   return (open?: boolean) => setExpanded(open ? panel : false)
  // }
  //   render(
  //     <>
  //     <Accordion expanded={expanded === 'panel1'} onClick={handleChange('panel1')}>
  //     <AccordionHeader >
  //       <p>{headerSec}</p>
  //     </AccordionHeader>
  //     <AccordionDetail>
  //       <p>{detailSec}</p>
  //     </AccordionDetail>
  //   </Accordion>
  //   <Accordion expanded={expanded === 'panel2'} onClick={handleChange('panel2')}>
  //   <AccordionHeader >
  //     <p>{headerSec2}</p>
  //   </AccordionHeader>
  //   <AccordionDetail>
  //     <p>{detailSec2}</p>
  //   </AccordionDetail>
  // </Accordion>
  // </>)
  //   const headerEl = screen.queryByText(headerSec) as HTMLDivElement
  //   const headerEl2 = screen.queryByText(headerSec2) as HTMLDivElement
  //   fireEvent.click(headerEl)
  //   fireEvent.click(headerEl2)
  //   const activeDiv = document.querySelectorAll('.accordion-container.active')
  //   expect(activeDiv).toHaveClass('accordion-container active')
  // })






})
