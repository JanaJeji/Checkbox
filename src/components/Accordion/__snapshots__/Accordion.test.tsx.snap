// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="accordion-container "
  >
    <div
      class="accordion-header "
    >
      <div
        class="accordion-header-btn"
      >
        <p>
          hello
        </p>
      </div>
    </div>
    <div
      class="accordion-detail-container"
      style="height: 0px;"
    >
      <div
        class="accordion-detail-container-content"
      >
        <p>
          New Accoirdon detail helloooo
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
